pkg/controller/association/dynamic_watches.go

```diff
@@ -78,11 +96,33 @@ func (r *Reconciler) setUserAndCaWatches(
 	return nil
 }
 
-func (r *Reconciler) removeWatches(associated types.NamespacedName) {
+func (r *Reconciler) removeWatches(associated types.NamespacedName, existing []commonv1.Association) {
 	// - ES resource
-	r.watches.ElasticsearchClusters.RemoveHandlerForKey(esWatchName(associated))
+	RemoveWatchesForDynamicRequest(associated, existing, esWatchNameRegexp, r.watches.ElasticsearchClusters)
 	// - ES CA Secret in the ES namespace
-	r.watches.Secrets.RemoveHandlerForKey(associatedCAWatchName(associated))
+	RemoveWatchesForDynamicRequest(associated, existing, associatedCAWatchNameRegexp, r.watches.Secrets)
 	// - user in the ES namespace
-	r.watches.Secrets.RemoveHandlerForKey(esUserWatchName(associated))
+	RemoveWatchesForDynamicRequest(associated, existing, esUserWatchNameRegexp, r.watches.Secrets)
+}
+
+func RemoveWatchesForDynamicRequest(
+	associated types.NamespacedName,
+	existing []commonv1.Association,
+	re *regexp.Regexp,
+	dynamicRequest *watches.DynamicEnqueueRequest,
+) {
+	lookup := make(map[string]bool)
+	for _, assoc := range existing {
+		lookup[fmt.Sprintf("%d", assoc.ID())] = true
+	}
+
+	for _, key := range dynamicRequest.Registrations() {
+		matches := re.FindStringSubmatch(key)
+		if len(matches) == 4 &&
+			matches[1] == associated.Namespace &&
+			matches[2] == associated.Name &&
+			!lookup[matches[3]] {
+			dynamicRequest.RemoveHandlerForKey(key)
```

Comment:
Do we need the regexes? Could we not use name mangling and structure the keys for easy retrieval idk something like `agent-es.default.my-agent.1.ca` as in `<assoc-type>.<namespace>.<name>.<seqno>.<watched-resource>` or something along those lines. Then deleting all watches for a given associated object would just be a prefix check on `agent-es.default.my-agent` ?

Reply:
Two points:

1. This also needs to be able to delete some (not all) of the watches for a resource, so somehow we need to recognise the id in the name in any case.
2. We could do prefix check for names:
`agent-es.default.my-agent.ca.1`
`agent-es.default.my-agent.es-user.1`
With:
`agent-es.default.my-agent.ca.`
`agent-es.default.my-agent.es-user.`

But this breaks depending on Agent name, eg. for Agent named `my-agent.ca.2`, the name would be:
`agent-es.default.my-agent.ca.2.ca.1` and our prefix match would see `2.ca.1`, then either fail or infer the id incorrectly.

While I don't think this could be a common issue, I'd be for being able to always match correctly rather than getting rid of regex to simplify the code slightly.